### ISSUE TITLE
mask-ratio not working when export onnx

### ISSUE THREAD

**Issue by UNeedCryDear** [ultralytics/yolov5#11948]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### Bug

train segment with mask-ratio=1,but when export to onnx, mask-ratio always is 4.

### Environment

Yolov5-seg: latest
torch:1.8.2-LST 
python: 3.8.15
cuda:11.8
cudnn:8.8
system: win10 21h2

### Minimal Reproducible Example

```
#train:
python segment/val.py --weights yolov5s-seg.pt --mask-ratio 1
# export:
python export.py --weights runs/train-seg/exp/weights/best.pt   --imgs [640,640] --opset 12 --include onnx 

```
Open onnx file using netron: https://netron.app/
input img-size is 640x640,and the protos(output1) shape is 1x32x160x160. After [process_mask_native() ](https://github.com/ultralytics/yolov5/blob/08974153767a3ab28b8be38c6ae203230518377d/utils/segment/general.py#L81)

mask is [n,160,160],which means that the mask for each instance shape is 1x160x160. 
The mask-ratio here is 4 instead of 1 during training.
![image](https://github.com/ultralytics/yolov5/assets/52729998/55cbfee1-05ff-4cf6-80d9-40514c082338)

### Additional

When inferring, can the parameters be set to be the same as during training, instead of being fixed to 4 times?

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@UNeedCryDear hi,

Thank you for reporting this issue. We apologize for any inconvenience caused.

The `mask-ratio` parameter is indeed not currently considered during ONNX export. It defaults to a fixed value of 4 in the exported ONNX model. We appreciate you bringing this to our attention.

We understand that it would be helpful to have the `mask-ratio` parameter customizable in the exported ONNX model. We encourage you to open a pull request with your proposed changes to address this issue. Our team and the YOLOv5 community would greatly appreciate your contribution.

Thank you again for your feedback and support.

Best regards,
Glenn

**Comment by UNeedCryDear** [comment]
It also defaults to a fixed value of 4 during the predicting. And I haven't found the code that where was the 4x defined.
Can you tell me where to find the  "fixed value of 4" definition is,so that I can have a try.

**Comment by glenn-jocher** [comment]
@UNeedCryDear thank you for the additional information.

The fixed value of 4 for the `mask-ratio` during inference can be found in the `utils/segment/general.py` file. In the `process_mask_native()` function, you will see a line where `mask_ratio` is set to 4.

You can try modifying this line in the code to use your desired value for `mask-ratio` during inference. However, please note that changing this value may impact the performance or accuracy of the segmentation results.

If you have any further questions or need assistance, feel free to ask.

Glenn

**Comment by UNeedCryDear** [comment]
@glenn-jocher 
Thank you for your information.
However, what I mean is that I want to set this magnification in the network, rather than resize it during post-processing. I hope that when outputs from the network. The shape of the protos is the size of the original image if mask_ratio is 1x. For example, if the input image is 1x3x640x640, I would like the shape of protos to be 1x32x640x640 instead of 1x32x160x160 now. It means the input shape is [1, 3, H, W], and the shape of  protos is [1, 32, H * mask_ratio, W * mask_ratio]

**Comment by glenn-jocher** [comment]
@UNeedCryDear

Thank you for your clarification.

I understand that you would like to set the magnification directly in the network, rather than resizing it during post-processing. Specifically, you would like the shape of the protos to match the size of the original image when the `mask_ratio` is set to 1x.

Currently, the YOLOv5 codebase does not support directly setting the `mask_ratio` in the network. The `mask_ratio` value is fixed to 4 in both the exported ONNX model and during inference.

To achieve the desired behavior, you would need to modify the code in the `utils/segment/general.py` file. You can adjust the dimensions of the `protos` tensor to match the size of the original image by multiplying the height and width by `mask_ratio`. However, please note that modifying this section of the code may require additional changes throughout the network.

If you have any further questions or need further assistance, feel free to let us know.

Glenn

**Comment by UNeedCryDear** [comment]
Thank you for your reply. I will give it a try.

**Comment by glenn-jocher** [comment]
@UNeedCryDear you're welcome! Feel free to give it a try and let us know if you have any further questions or need any assistance. We're here to help.

**Comment by arnaud-nt2i** [comment]
@glenn-jocher,  @UNeedCryDear 
There is something I still don't understand, when exporting model to onnx, there is no use of utils/segment/general.py
And protos output is [1, 32, 160,160] (with input 640) so it seems to me this 4x ratio is hard coded in the model.

But if so, what is the purpose of mask_ratio ?
Mask are learned in variable quality (mask_ratio) but always predicted with a 4x reduction factor ?

I add that I need really detailed segmentation for my application and I am using custom NMS.

**Comment by glenn-jocher** [comment]
@arnaud-nt2i the purpose of the `mask_ratio` parameter is to control the downsampling factor applied to the predicted masks during inference. While the masks are trained at a higher resolution to capture more details (`mask_ratio` > 1), they are downsampled by a fixed factor of 4 during inference for computational efficiency. This downsampling helps to reduce the memory and computational requirements during inference.

We understand that for some applications, a more detailed segmentation is desired. In that case, you can modify the YOLOv5 code to adjust the downsampling factor, either by modifying the `process_mask_native()` function in `utils/segment/general.py` or by modifying the corresponding part in the YOLOv5 model architecture.

In addition, using custom NMS can help improve the segmentation output by fine-tuning the object merging and refinement process.

If you have any further questions or need assistance with your application, please let us know.

Glenn
